cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -345,9 +390,23 @@ private void processMetrics() {
           // Set persistingFlag back to false after persisting completes.
           persistingFlag.set(false);
         }
+        if (recordsCopy.size() == 0) {
+          // Should never reach here;
+          return metricsProcessIntervalMillis;
+        }
+        // Calculating the length of time used to emit the metrics to be persisted.
+        long metricsEmittingTime = recordsCopy.peekLast().getTimestamp() - recordsCopy.peekFirst().getTimestamp();
+        // Estimate number of metrics emitted to be persisted per millisecond
+        double metricsEmittingSpeed = newMetricsCount / metricsEmittingTime;
```

Comment:
My theory is that `(process time + sleep time)*metricsEmittingSpeed = N`, where N is number of metrics to be processed in the next run. If N > fetcherLimit, unprocessed metrics will accumulate after each new run, since at most fetcherLimit metrics can be processed at each run. Therefore, we should determine the sleep time by the condition N < fetcherLimit